media: i2c: imx477: Add support for imx378 as a compatible sensor

davidplowman opened the pull request:
The imx378 is almost identical to the imx477 and can be supported by
the same driver with just a few extra register writes.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

davidplowman commented:
@naushir @6by9 
Here's a new version of this. It adds the imx378 as a "compatible" sensor to the imx477. You need to specify the overlay as either "dtoverlay=imx477" or "dtoverlay=imx378", and it has to match the attached sensor correctly.

One thing I was a bit unsure of was how to update the documentation, particularly as these two sensors are both different, yet intertwined. Any advice there?

6by9 commented on drivers/media/i2c/imx477.c:
```diff
@@ -1673,11 +1683,18 @@ static int imx477_start_streaming(struct imx477 *imx477)
 	if (!imx477->common_regs_written) {
 		ret = imx477_write_regs(imx477, mode_common_regs,
 					ARRAY_SIZE(mode_common_regs));
+		if (ret == 0) {
```
```if (!ret)```

davidplowman replied:
Okeydoke!

6by9 commented on drivers/media/i2c/imx477.c:
```diff
@@ -2090,6 +2138,12 @@ static int imx477_probe(struct i2c_client *client)
+	if (!match)
+		return -ENODEV;
+	memcpy(&imx477->compatible_data, match->data,
+	       sizeof(imx477->compatible_data));
```
Why memcpy instead of having ```struct imx477_compatible_data *compatible_data;``` in the struct and copy the pointer?

davidplowman replied:
True enough, will fix that!

6by9 reviewed (commented):
A couple of minor comments on the code.

Ideally dt-bindings, dtoverlays, and driver changes should be split into independent commits.

davidplowman commented:
So just to try and get this right, should that be 2 commits, one with the driver (presumably that's the 2nd of the two) and one with everything else?  Thanks!

6by9 commented:
> So just to try and get this right, should that be 2 commits, one with the driver (presumably that's the 2nd of the two) and one with everything else? Thanks!

3 please.
- Documentation/devicetree
- arch/arm/boot/dts/boot/overlays
- drivers/media/i2c
It's the normal breakdown for mainline as they are different contributing trees.
boot/dts/overlays is only in our downstream tree, but it is generally squashed when switching to a new release, and that's made more awkward if you've mixed overlays with other stuff.

Can I check that you're running ```scripts/checkpatch.pl --strict``` against your patches before submission? I have a .git/hooks/post-commit hook of 
```
exec git show --format=email HEAD | ./scripts/checkpatch.pl --strict --codespell
```
so that is automatic on every commit.

davidplowman committed:
Documentation: devicetree: Add documentation for imx378 sensor

The imx378 sensor is compatible with the imx477 and shares common
device tree settings.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

davidplowman committed:
overlays: Add overlay for imx378 sensor

This is based off a common overlay which is now also used by the
imx477 sensor.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

davidplowman committed:
media: i2c: imx477: Extend driver to support imx378 sensor

The imx378 sensor is almost identical to the imx477 and can be
supported as a "compatible" sensor with just a few extra register
writes.

Signed-off-by: David Plowman <david.plowman@raspberrypi.com>

pelwell commented on arch/arm/boot/dts/overlays/imx378-overlay.dts:
```diff
@@ -0,0 +1,21 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
```
This header isn't required.

pelwell commented on arch/arm/boot/dts/overlays/imx477-overlay.dts:
```diff
@@ -4,110 +4,18 @@
 /plugin/;
 
 #include <dt-bindings/gpio/gpio.h>
```
Nor here.

pelwell commented on arch/arm/boot/dts/overlays/imx477_378-overlay.dtsi:
```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Definitions for IMX477 camera module on VC I2C bus
+/dts-v1/;
```
This line and the next should only appear once - at the top of the including .dts files.

pelwell commented on arch/arm/boot/dts/overlays/imx477_378-overlay.dtsi:
```diff
@@ -0,0 +1,112 @@
+/dts-v1/;
+/plugin/;
+
+#include <dt-bindings/gpio/gpio.h>
```
Still not needed.

6by9 commented:
It tells you the file and line:
```
WARNING: Possible repeated word: 'git'
#143: FILE: MAINTAINERS:16362:
+T:	git git://linuxtv.org/media_tree.git
```
Line 16362 is correct, except that I believe you now have drivers/media/i2c/imx477.c listed twice in MAINTAINERS, which is a bad thing.
imx378.yaml can be listed under the one entry for IMX477 at present.
I'm not sure what upstream will truly want in this case. You could list the one binding with multiple compatibles, but this is more generic should they ever need to diverge.

```WARNING: added, moved or deleted file(s), does MAINTAINERS need updating?```
You added arch/arm/boot/dts/overlays/imx378-overlay.dts and renamed /arch/arm/boot/dts/overlays/imx477_378-overlay.dtsi
None of the overlays are covered by MAINTAINTERS, so no issue.

pelwell commented on arch/arm/boot/dts/overlays/imx477-overlay.dts:
```diff
@@ -4,110 +4,18 @@
+#include "imx477_378-overlay.dtsi"
 
 /{
 	compatible = "brcm,bcm2835";
```
As with the imx378 overlay, the remainder of the file can just be:
```
&imx477 {
        compatible = "sony,imx477";
};
```

pelwell commented on arch/arm/boot/dts/overlays/imx378-overlay.dts:
```diff
@@ -0,0 +1,21 @@
+#include <dt-bindings/gpio/gpio.h>
+#include "imx477_378-overlay.dtsi"
+
+/{
```
There's no need to repeat the compatible string. In fact, because the change of compatible string is unconditional, this whole node (the remainder of the file) can be replaced with:
```
&imx477 {
        compatible = "sony,imx378";
};
```

davidplowman commented:
Thanks Phil, Dave. Will look again at those various bits and resubmit!

davidplowman commented:
Latest attempt, also down to just the one warning now... hopefully we're getting a bit closer!  :)

6by9 commented:
LGTM.
